hail/src/main/scala/is/hail/expr/ir/functions/StringFunctions.scala

```diff
@@ -51,12 +51,13 @@ object StringFunctions extends RegistryFunctions {
   def registerAll(): Unit = {
     val thisClass = getClass
 
-    registerCode("length", TString(), TInt32()) { (r: EmitRegion, s: Code[Long]) =>
-      asm4s.coerce[String](wrapArg(r, TString())(s)).invoke[Int]("length")
+    registerCode("length", TString(), TInt32()) { case (r: EmitRegion, (sT: PString, s: Code[Long])) =>
+      asm4s.coerce[String](wrapArg(r, sT)(s)).invoke[Int]("length")
     }
 
-    registerCode("slice", TString(), TInt32(), TInt32(), TString()) { (r: EmitRegion, s: Code[Long], start: Code[Int], end: Code[Int]) =>
-      unwrapReturn(r, TString())(asm4s.coerce[String](wrapArg(r, TString())(s)).invoke[Int, Int, String]("substring", start, end))
+    registerCode("slice", TString(), TInt32(), TInt32(), TString()) {
+      case (r: EmitRegion, (sT: PString, s: Code[Long]), (startT, start: Code[Int]), (endT, end: Code[Int])) =>
```

Comment:
To confirm: startT and endT should be equal correct? We don't use the endT, so there seems to be a simpler way to represent this. Also, should we check that startT == endT? , or create a version of registerCode that forces this constraint, e.g `registerCode[T1]`?

Reply:
hmm, after thinking about it more, I think this is probably wrong. I think PInterval needs to parameterize both the start ptype and end ptype, but that's out-of-scope for this change.

For now I'll change this to use `PType.canonical(startT.virtualType)`, and we do know that the virtual type is the same between startT and endT.